vercel/turbo
# Add `examples/with-docker-pnpm` sample of using pnpm and docker

### Description
Based on:
* https://github.com/vercel/turbo/issues/5462 
* https://github.com/vercel/turbo/issues/534

This PR adds a new example of using docker pnpm.

I've tried to keep the project code and structure similar to the existing `examples/with-docker` project with some updated goals:
* The `api` app is **strongly typed**, but uses standard ESM instead of relying on a build step. Additional type information can be added using JsDoc. 
* The Dockerfile might not make the best performance decision (I assume), but I've kept it simple enough so that I could understand it. Initially I didn't even use the node alpine version, but switched to it afterwards.
* I've removed most next config configurations and left just these:
```
export default {
  reactStrictMode: true,
  experimental: {
    appDir: true,
  },
};
```

I've also installed `syncpack` to check if there are any mismatches. I've added a package.json script for it, and it can be ran like this:
```
pnpm run syncpack
```
Unfortunately it also show errors for the local dependencies: https://github.com/JamieMason/syncpack/issues/121

but it will also show other mismatches, so I could pretty easily fix those.

### Testing Instructions
I've personally tested this by
1. cloning the repo
2. cd into `./examples/with-docker-pnpm`
3. Follow the instruction in the readme (`pnpm install`)

The previous step may not work by default (it didn't for me) because the root package.json has this:
```
  "engines": {
    "node": "18.x",
    "pnpm": "7"
  }
```
And my local pnpm was `8.6.5`.
While writing this PR I just removed that requirement for pnpm.

This doesn't seem like a nice solution.

A simpler solution would be to copy the `with-docker-pnpm` folder into a different folder and follow the instruction in the README. This seems like a better solution since this is how a potential user would use the example.

After copying the folder, run:
```
pnpm install
pnpm run dev
```

and/or

```
docker compose build
docker compose up
```

 both methods should work.
